bindings/profilers/wall.cc

```diff
@@ -25,213 +28,609 @@
-                        Nan::New<Integer>(node->GetHitCount()),
-                        children);
-}
+  Local<Value> TranslateTimeProfileNode(const CpuProfileNode* node) {
+    int32_t count = node->GetChildrenCount();
+    Local<Array> children = Nan::New<Array>(count);
+    for (int32_t i = 0; i < count; i++) {
+      Nan::Set(children, i, TranslateTimeProfileNode(node->GetChild(i)));
+    }
 
-Local<Value> TranslateTimeProfile(const CpuProfile* profile,
-                                  bool includeLineInfo) {
-  Local<Object> js_profile = Nan::New<Object>();
-  Nan::Set(js_profile,
-           Nan::New<String>("title").ToLocalChecked(),
-           profile->GetTitle());
+    uint32_t hitcount = 0;
+    auto labels = getLabelSetsForNode(node, hitcount);
+
+    return CreateTimeNode(node->GetFunctionName(),
+                          node->GetScriptResourceName(),
+                          NewInteger(node->GetScriptId()),
+                          NewInteger(node->GetLineNumber()),
+                          NewInteger(node->GetColumnNumber()),
+                          NewInteger(hitcount),
+                          children,
+                          labels);
+  }
 
-#if NODE_MODULE_VERSION > NODE_11_0_MODULE_VERSION
-  if (includeLineInfo) {
+  Local<Value> TranslateTimeProfile(const CpuProfile* profile,
+                                    bool includeLineInfo) {
+    Local<Object> js_profile = Nan::New<Object>();
+
+    if (includeLineInfo) {
+      Nan::Set(js_profile,
+               NewString("topDownRoot"),
+               TranslateLineNumbersTimeProfileRoot(profile->GetTopDownRoot()));
+    } else {
+      Nan::Set(js_profile,
+               NewString("topDownRoot"),
+               TranslateTimeProfileNode(profile->GetTopDownRoot()));
+    }
     Nan::Set(js_profile,
-             Nan::New<String>("topDownRoot").ToLocalChecked(),
-             TranslateLineNumbersTimeProfileRoot(profile->GetTopDownRoot()));
-  } else {
+             NewString("startTime"),
+             Nan::New<Number>(profile->GetStartTime()));
     Nan::Set(js_profile,
-             Nan::New<String>("topDownRoot").ToLocalChecked(),
-             TranslateTimeProfileNode(profile->GetTopDownRoot()));
+             NewString("endTime"),
+             Nan::New<Number>(profile->GetEndTime()));
+
+    return js_profile;
   }
-#else
-  Nan::Set(js_profile,
-           Nan::New<String>("topDownRoot").ToLocalChecked(),
-           TranslateTimeProfileNode(profile->GetTopDownRoot()));
-#endif
-  Nan::Set(js_profile,
-           Nan::New<String>("startTime").ToLocalChecked(),
-           Nan::New<Number>(profile->GetStartTime()));
-  Nan::Set(js_profile,
-           Nan::New<String>("endTime").ToLocalChecked(),
-           Nan::New<Number>(profile->GetEndTime()));
-  return js_profile;
-}
+};
 
-WallProfiler::WallProfiler(int interval) : samplingInterval(interval) {}
+LabelSetsByNode WallProfiler::GetLabelSetsByNode(CpuProfile* profile,
```

Comment:
Should we get a metric out of the function on the number of labeled samples ? How can we know if this works well when it runs in staging ?

Reply:
Only labelled samples ends up in the emitted profile, so we should indirectly see if it's working

Reply:
:+1: This makes it easy for long lived applications, though not easy for shorter applications (if we don't know how long the application ran for).